File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/RAG/ViewModels/RAGViewModel.swift
//
//  RAGViewModel.swift
//  RunAnywhereAI
//
//  ViewModel for the RAG feature. Orchestrates document loading,
//  text extraction, SDK pipeline lifecycle, and query flow.
//

import Foundation
import Observation
import RunAnywhere
import os.log

// MARK: - Message Role

enum MessageRole {
    case user
    case assistant
    case system
}

// MARK: - RAG View Model

@MainActor
@Observable
final class RAGViewModel {

    // MARK: - Document State

    private(set) var documentName: String?
    private(set) var isDocumentLoaded = false
    private(set) var isLoadingDocument = false

    // MARK: - Query State

    private(set) var messages: [(role: MessageRole, text: String)] = []
    private(set) var isQuerying = false
    /// Settable from the view layer to surface file-picker failures in the error banner.
    var error: Error?

    // MARK: - Input

    var currentQuestion = ""

    // MARK: - Private

    private let logger = Logger(subsystem: "com.runanywhere.RunAnywhereAI", category: "RAGViewModel")

    // MARK: - Computed Properties

    var canAskQuestion: Bool {
        isDocumentLoaded
        && !isQuerying
        && !currentQuestion.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
    }

    // MARK: - Public Methods

    /// Load a document: extract text, create RAG pipeline, ingest text.
    ///
    /// - Parameters:
    ///   - url: Security-scoped URL of the document (PDF or JSON).
    ///   - config: RAG pipeline configuration with model paths and tuning parameters.
    func loadDocument(url: URL, config: RAGConfiguration) async {
        isLoadingDocument = true
        error = nil

        defer {
            isLoadingDocument = false
        }

        do {
            logger.info("Extracting text from document: \(url.lastPathComponent)")
            let extractedText = try DocumentService.extractText(from: url)

            logger.info("Creating RAG pipeline")
            try await RunAnywhere.ragCreatePipeline(config: config)

            logger.info("Ingesting document text (\(extractedText.count) chars)")
            try await RunAnywhere.ragIngest(text: extractedText)

            documentName = url.lastPathComponent
            isDocumentLoaded = true
            logger.info("Document loaded successfully: \(url.lastPathComponent)")
        } catch {
            self.error = error
            logger.error("Failed to load document: \(error.localizedDescription)")
        }
    }

    /// Query the loaded document with the current question.
    ///
    /// Appends the user question and the assistant answer to `messages`.
    /// Guards against empty questions and unloaded documents.
    func askQuestion() async {
        let question = currentQuestion.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !question.isEmpty else { return }
        guard isDocumentLoaded else { return }

        messages.append((role: .user, text: question))
        currentQuestion = ""
        isQuerying = true
        error = nil

        defer {
            isQuerying = false
        }

        do {
            logger.info("Querying RAG pipeline: \(question)")
            let result = try await RunAnywhere.ragQuery(question: question)
            messages.append((role: .assistant, text: result.answer))
            logger.info("Query complete (\(result.totalTimeMs, format: .fixed(precision: 0))ms)")
        } catch {
            self.error = error
            messages.append((role: .assistant, text: "Error: \(error.localizedDescription)"))
            logger.error("Query failed: \(error.localizedDescription)")
        }
    }

    /// Clear the loaded document and destroy the RAG pipeline.
    ///
    /// Resets all document and conversation state.
    func clearDocument() async {
        await RunAnywhere.ragDestroyPipeline()

        documentName = nil
        isDocumentLoaded = false
        messages = []
        error = nil
        currentQuestion = ""

        logger.info("Document cleared and pipeline destroyed")
    }
}

File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/rag/RAGViewModel.kt
package com.runanywhere.runanywhereai.presentation.rag

import android.content.Context
import android.net.Uri
import timber.log.Timber
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.runanywhere.runanywhereai.domain.services.DocumentService
import com.runanywhere.runanywhereai.domain.services.DocumentServiceError
import com.runanywhere.sdk.public.RunAnywhere
import com.runanywhere.sdk.public.extensions.RAG.RAGConfiguration
import com.runanywhere.sdk.public.extensions.ragCreatePipeline
import com.runanywhere.sdk.public.extensions.ragDestroyPipeline
import com.runanywhere.sdk.public.extensions.ragIngest
import com.runanywhere.sdk.public.extensions.ragQuery
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

// MARK: - Message Role

/**
 * Role of a RAG conversation message.
 *
 * Scoped to the RAG feature to avoid polluting shared AppTypes.
 * Mirrors iOS MessageRole enum exactly.
 */
enum class RAGMessageRole {
    USER,
    ASSISTANT,
    SYSTEM,
}

// MARK: - RAG Message

/**
 * A single message in the RAG conversation.
 *
 * Uses a structured data class instead of a tuple (iOS uses anonymous tuples).
 * Mirrors iOS `(role: MessageRole, text: String)` pattern with a named type.
 */
data class RAGMessage(
    val role: RAGMessageRole,
    val text: String,
)

// MARK: - RAG UI State

/**
 * Immutable UI state for the RAG screen.
 *
 * Mirrors iOS RAGViewModel published properties as a single consolidated state.
 */
data class RAGUiState(
    /** Display name of the loaded document (last path component). */
    val documentName: String? = null,

    /** Whether a document has been fully ingested into the pipeline. */
    val isDocumentLoaded: Boolean = false,

    /** Whether the document extraction + pipeline creation is in progress. */
    val isLoadingDocument: Boolean = false,

    /** All conversation messages (user questions and assistant answers). */
    val messages: List<RAGMessage> = emptyList(),

    /** Whether a query is currently running. */
    val isQuerying: Boolean = false,

    /** Last error message, if any. */
    val error: String? = null,

    /** Current text entered by the user in the question field. */
    val currentQuestion: String = "",
) {
    /**
     * Whether the user can submit a question.
     * Mirrors iOS `canAskQuestion` computed property.
     */
    val canAskQuestion: Boolean
        get() = isDocumentLoaded && !isQuerying && currentQuestion.isNotBlank()
}

// MARK: - RAG ViewModel

/**
 * ViewModel for the RAG feature.
 *
 * Orchestrates document loading (text extraction, pipeline creation, ingestion),
 * query flow (vector search + LLM generation), and pipeline teardown.
 *
 * Mirrors iOS RAGViewModel exactly, adapted for Android ViewModel + StateFlow + viewModelScope.
 */
class RAGViewModel : ViewModel() {

    private val _uiState = MutableStateFlow(RAGUiState())
    val uiState: StateFlow<RAGUiState> = _uiState.asStateFlow()

    // MARK: - Input

    /**
     * Update the current question text typed by the user.
     * Call this from the Compose TextField's onValueChange callback.
     */
    fun updateQuestion(text: String) {
        _uiState.update { it.copy(currentQuestion = text) }
    }

    // MARK: - Pipeline Lifecycle

    /**
     * Load a document: extract text, create RAG pipeline, ingest text.
     *
     * Mirrors iOS `RAGViewModel.loadDocument(url:config:)`.
     *
     * @param context Android context for ContentResolver access
     * @param uri Content URI of the document (PDF or JSON)
     * @param config RAG pipeline configuration with model paths and tuning parameters
     */
    fun loadDocument(context: Context, uri: Uri, config: RAGConfiguration) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoadingDocument = true, error = null) }

            try {
                val fileName = DocumentService.getFileName(context, uri) ?: "Document"
                Timber.i("Extracting text from document: $fileName")
                val extractedText = withContext(Dispatchers.IO) {
                    DocumentService.extractText(context, uri)
                }

                Timber.i("Creating RAG pipeline")
                RunAnywhere.ragCreatePipeline(config)

                Timber.i("Ingesting document text (${extractedText.length} chars)")
                RunAnywhere.ragIngest(text = extractedText)

                _uiState.update {
                    it.copy(
                        documentName = fileName,
                        isDocumentLoaded = true,
                    )
                }
                Timber.i("Document loaded successfully: $fileName")
            } catch (e: DocumentServiceError) {
                Timber.e("Document extraction failed: ${e.message}")
                _uiState.update { it.copy(error = e.message) }
            } catch (e: Exception) {
                Timber.e(e, "Failed to load document: ${e.message}")
                _uiState.update { it.copy(error = e.message ?: "Failed to load document") }
            } finally {
                _uiState.update { it.copy(isLoadingDocument = false) }
            }
        }
    }

    /**
     * Query the loaded document with the current question.
     *
     * Appends the user question and the assistant answer (with timing) to messages.
     * Guards against empty questions and unloaded documents.
     *
     * Mirrors iOS `RAGViewModel.askQuestion()`.
     */
    fun askQuestion() {
        val question = _uiState.value.currentQuestion.trim()
        if (question.isEmpty() || !_uiState.value.isDocumentLoaded) return

        viewModelScope.launch {
            // Append user message, clear input, mark querying
            _uiState.update {
                it.copy(
                    messages = it.messages + RAGMessage(role = RAGMessageRole.USER, text = question),
                    currentQuestion = "",
                    isQuerying = true,
                    error = null,
                )
            }

            try {
                Timber.i("Querying RAG pipeline: $question")
                val result = RunAnywhere.ragQuery(question = question)

                val answerWithTiming = "${result.answer}\n\nAnswer generated in ${
                    String.format("%.1f", result.totalTimeMs / 1000.0)
                }s"

                _uiState.update {
                    it.copy(
                        messages = it.messages + RAGMessage(
                            role = RAGMessageRole.ASSISTANT,
                            text = answerWithTiming,
                        ),
                    )
                }
                Timber.i("Query complete (${result.totalTimeMs}ms)")
            } catch (e: Exception) {
                Timber.e(e, "Query failed: ${e.message}")
                val errorText = "Error: ${e.message ?: "Query failed"}"
                _uiState.update {
                    it.copy(
                        messages = it.messages + RAGMessage(
                            role = RAGMessageRole.ASSISTANT,
                            text = errorText,
                        ),
                        error = e.message,
                    )
                }
            } finally {
                _uiState.update { it.copy(isQuerying = false) }
            }
        }
    }

    /**
     * Clear the loaded document and destroy the RAG pipeline.
     *
     * Resets all document and conversation state.
     * Mirrors iOS `RAGViewModel.clearDocument()`.
     */
    fun clearDocument() {
        viewModelScope.launch {
            try {
                RunAnywhere.ragDestroyPipeline()
                Timber.i("Document cleared and pipeline destroyed")
            } catch (e: Exception) {
                Timber.e(e, "Failed to destroy pipeline: ${e.message}")
            } finally {
                _uiState.update {
                    RAGUiState() // Reset to initial state
                }
            }
        }
    }
}
